web_autoflip_demo/autoflip_demo_UI/autoflip_worker.ts

```diff
@@ -0,0 +1,340 @@
+      // Sets the aspect ratio setAspectRatio(1, 1) = '1:1' or
+      // setAspectRatio(16, 11) = '16: 11'.
+      autoflipModule.setAspectRatio(16, 8);
+    });
+});
+
+/** Analyzes the input frames and output caculated crop windows for each frame. */
+onmessage = function (e: MessageEvent): void {
+  let signal: Signal = e.data;
+  console.log(`AUTOFLIP: video(${signal.videoId}) start to crop`, signal);
+  if (signal.type === 'firstCrop') {
+    videoWidth = signal.width;
+    videoHeight = signal.height;
+    workerWindow = signal.window;
+  }
+  // Gets frameData from indexDB and process with Autoflip.
+  read(signal.videoId, signal.startTime).then((value): void => {
+    console.log(`PROMISE: promise ${signal.videoId} returned`);
+    let frameData: Frame[] = value;
+    handleFrames(frameData, signal);
+    if (signal.end === true) {
+      const b = autoflipModule.closeGraphInternal();
+      // This posts the analysis result back to main script.
+      ctx.postMessage({
+        type: 'finishedAnalysis',
+        cropWindows: resultCropInfo,
+        startId: currentId,
+        videoId: signal.videoId,
+        shots: resultShots,
+      });
+    } else {
+      ctx.postMessage({
+        type: 'nextCrop',
+        videoId: signal.videoId + 1,
+      });
+    }
+  });
+};
+
+/** Processes input frames with autoflip wasm. */
+function handleFrames(frameData: Frame[], signal: Signal): void {
+  console.log(
+    `AUTOFLIP: video (${signal.videoId}) received from main`,
+    frameData,
+  );
+  const info = { width: videoWidth, height: videoHeight };
+  for (let i = 0; i < frameData.length; i++) {
+    const image = frameData[i].data; // Array buffer from FFmpeg, .tiff <-
+    // Saving array buffer to the wasm heap memory.
+    const numBytes = image.byteLength;
+    const ptr = ctx.Module._malloc(numBytes);
+    const heapBytes = new Uint8Array(ctx.Module.HEAPU8.buffer, ptr, numBytes);
+    const uint8Image = new Uint8Array(image);
+    heapBytes.set(uint8Image);
+    // End saving memory.
+    const status = autoflipModule.processRawYuvBytes(
+      ptr,
+      info.width,
+      info.height,
+    );
+    // Do this to check if it saved correctly.
+    if (!status) {
+      console.error(status);
+      throw new Error('Autoflip had an error!');
+    }
+    // Finish.
+    autoflipModule._free(ptr);
+  }
+
+  const a = autoflipModule.runTillIdle();
+
+  if (resultCropInfo.length !== 0) {
+    // This posts the analysis result back to main script.
+    ctx.postMessage({
+      type: 'currentAnalysis',
+      cropWindows: resultCropInfo,
+      startId: currentId,
+      videoId: signal.videoId,
+      shots: resultShots,
+    });
+    currentId += resultCropInfo.length;
+    resultCropInfo = [];
+    resultShots = [];
+  } else {
+    console.log(`AUTOFLIP: runTillIdle is not avaiable`);
+  }
+}
+
+/** Reads frame decode datarows from the indexDB. */
+async function read(videoId: number, start: number): Promise<Frame[]> {
```

Comment:
Maybe use a more descriptive function name, like readFramesFromIndexedDb?

Reply:
Totally agree, change some function names. Please indicate more if you see any name improper.

Reply:
Great work!